File: apps/game-ios/Sources/ALUI/Components/ListenPill.swift
import SwiftUI

/* -------------------------------------------------------------------------- */
/* The big 🔊 « Écouter » button, once — D55.                                   */
/*                                                                             */
/* Every exercise puts one under its mascot, and all eight TSX files author it  */
/* with the same class list:                                                   */
/*                                                                             */
/*   rounded-full bg-white/70 px-5 py-2 text-lg font-bold text-[#5A3A1E]       */
/*   shadow [touch-action:none]                                                */
/*                                                                             */
/* The Swift port grew four copies of that — `LettersListenPill`,              */
/* `SpellListenPill`, `SoundEngineChrome.listenButton` and `AssembleView`'s     */
/* private one — which is exactly the shape of divergence CLAUDE.md warns       */
/* about: eight identical lines of JSX, ported four times, so a fix lands in    */
/* one of them and nobody sees the other three. It is one component now, and    */
/* the four names above are thin wrappers that only differ in the margin they   */
/* carry.                                                                      */
/*                                                                             */
/* [DEVIATION, reported] It PRESSES. The web's does not: `Tile.tsx` is the only */
/* file that calls `el.animate`, so on the web this pill is visually inert and  */
/* the only acknowledgement of a tap is the voice that follows it. On a phone   */
/* that reads as a dead button — the clip may still be decoding, and a child    */
/* who gets nothing back taps again, which cuts the line they just asked for.   */
/* It now runs the same 130 ms squish a tile does, through the same             */
/* `TilePress.previewDown` path: press first, then speak, and never a shake     */
/* (asking to hear something has no verdict — invariant 3).                     */
/* -------------------------------------------------------------------------- */

/// The pill's authored metrics — the Tailwind classes, once. A separate enum
/// rather than statics on the view because the view is `@MainActor` and the
/// engines re-export these from their own (nonisolated) metric namespaces.
public enum ListenPillMetrics {
    /// `px-5`.
    public static let paddingX: CGFloat = 20
    /// `py-2`.
    public static let paddingY: CGFloat = 8
    /// Tailwind's plain `shadow` utility: `0 1px 3px rgba(0,0,0,0.1),
    /// 0 1px 2px -1px rgba(0,0,0,0.1)`.
    public static let shadow = CSSShadow(y: 1, blur: 3, opacity: 0.1)
    public static let shadowTight = CSSShadow(y: 1, blur: 2, opacity: 0.1)
}

@MainActor
struct ListenPill: View {

    /// What the button SHOWS — « 🔊 Écouter », or `🔊 {word}` where the engine
    /// prints the prompt on the button itself.
    let text: String
    let accessibilityLabel: String
    /// Runs at touch-down, after the press starts. The `locked` guard that stops
    /// it cutting a success line lives in the model's `replayPrompt()`.
    let action: () -> Void

    @State private var handle = LayerHandle()

    var body: some View {
        LayerHost(handle: handle) {
            Text(verbatim: text)
                .font(Typography.rounded(Typography.Size.lg, Typography.Weight.bold))
                .foregroundStyle(Palette.ink.color)
                .padding(.horizontal, ListenPillMetrics.paddingX)
                .padding(.vertical, ListenPillMetrics.paddingY)
                // The shadow is cast by the CAPSULE, inside the background —
                // never by the composed pill, whose glyphs would each cast one
                // of their own (D54).
                .background {
                    Capsule()
                        .fill(.white.opacity(Palette.White.o70))
                        .shadow(
                            color: .black.opacity(ListenPillMetrics.shadow.opacity),
                            radius: ListenPillMetrics.shadow.swiftUIRadius,
                            y: ListenPillMetrics.shadow.y)
                        .shadow(
                            color: .black.opacity(ListenPillMetrics.shadowTight.opacity),
                            radius: ListenPillMetrics.shadowTight.swiftUIRadius,
                            y: ListenPillMetrics.shadowTight.y)
                }
        }
        .contentShape(Capsule())
        .touchDown {
            // Invariant 1 — press and speak both happen inside the touch-down
            // call, before SwiftUI commits anything.
            TilePress.previewDown(layer: handle.layer, disabled: false, onPreview: action)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(accessibilityLabel)
        .accessibilityAddTraits(.isButton)
    }
}

File: apps/game-android/ui/src/main/kotlin/fr/dappit/attrapelettres/ui/components/ListenPill.kt
package fr.dappit.attrapelettres.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.onClick
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import fr.dappit.attrapelettres.ui.design.Palette
import fr.dappit.attrapelettres.ui.design.Typography
import fr.dappit.attrapelettres.ui.interaction.rememberTileMotion
import fr.dappit.attrapelettres.ui.interaction.tileMotion
import fr.dappit.attrapelettres.ui.interaction.touchDown

// The big « Ecouter » button, once — the twin of iOS's `ListenPill.swift` (D55).
//
// Every exercise puts one under its mascot, and all eight TSX files author it
// with the same class list:
//
//   rounded-full bg-white/70 px-5 py-2 text-lg font-bold text-[#5A3A1E]
//   shadow [touch-action:none]
//
// The Swift port grew FOUR copies of that before anybody noticed, which is
// exactly the divergence the root CLAUDE.md warns about: eight identical lines
// of JSX ported four times, so a fix lands in one of them and nobody sees the
// other three. It is one composable here from the start, and an engine that
// paints its own white capsule is a bug, not a variation.
//
// It PRESSES — the deviation iOS reported and kept. The web's does not:
// `Tile.tsx` is the only file that calls `el.animate`, so on the web this pill
// is visually inert and the only acknowledgement of a tap is the voice that
// follows it. On a phone that reads as a dead button — the clip may still be
// warming, and a child who gets nothing back taps again, which cuts the line
// they just asked for. It runs the same 130 ms squish a tile does, through the
// same `TileMotion.press()`, and never a shake: asking to hear something has no
// verdict (invariant 3).
//
// INVARIANT 1 lives in the `touchDown` below. The web fires this button on
// `onPointerDown`, not `onClick`, and so do we: the press and the call to
// [onListen] both happen inside the pointer dispatch, before anything commits.
// The `locked` guard that stops a replay cutting a success line belongs to the
// engine, not here.

/**
 * Tailwind `text-lg` sets a line-height of 28 px alongside its 18 px font size,
 * and both places that wear the class here — this pill and « ← Menu » — depend
 * on it: 28 + 8 + 8 = 44 dp of button, which is how the web's own arithmetic
 * clears the 44 pt tap floor without a single explicit height.
 *
 * `Typography.LineHeight` carries the `leading-*` utilities; this is the ratio
 * baked into the SIZE utility, so it lives beside the two call sites.
 */
const val TEXT_LG_LEADING: Float = 28f / 18f

/** The pill's authored metrics — the Tailwind classes, once. */
object ListenPillMetrics {

    /** `px-5`. */
    val paddingX: Dp = 20.dp

    /** `py-2`. */
    val paddingY: Dp = 8.dp

    /** [TEXT_LG_LEADING] — `text-lg`'s own 28 px line box. */
    const val textRatio: Float = TEXT_LG_LEADING

    /** Tailwind's plain `shadow` utility, both layers. */
    val shadow = Shadows.tailwind

    /**
     * The resulting height: `py-2` twice plus the 28 px line box.
     *
     * Recorded rather than enforced with a `heightIn` — the web's own arithmetic
     * already clears 44 dp, and pinning it in a test is how we find out if a
     * typography change ever stops it (invariant 6).
     */
    val height: Dp = 44.dp
}

/**
 * The « Ecouter » button.
 *
 * @param text what the button SHOWS — `Copy.Exercise.LISTEN`, or
 *   `Copy.Exercise.listenWord(word)` where the engine prints the prompt on the
 *   button itself.
 * @param contentDescription what a screen reader announces — the per-engine
 *   wording from `Copy.Exercise` (« Répéter le mot », « Réécouter le son », …).
 * @param onListen runs at touch-DOWN, right after the press starts.
 */
@Composable
fun ListenPill(
    text: String,
    contentDescription: String,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    onListen: () -> Unit,
) {
    val motion = rememberTileMotion()
    val fontScale = LocalDensity.current.fontScale
    Box(
        modifier
            // One accessible element carrying the engine's wording, plus a real
            // click ACTION so TalkBack's double-tap activates it: `touchDown`
            // deliberately adds no semantics of its own, and an element with no
            // click action is one an assistive gesture has to fall through to.
            .semantics(mergeDescendants = true) {
                role = Role.Button
                this.contentDescription = contentDescription
                onClick {
                    if (enabled) onListen()
                    true
                }
            }
            .tileMotion(motion)
            .cssShadow(ListenPillMetrics.shadow, CircleShape)
            .background(Color.White.copy(alpha = Palette.White.o70), CircleShape)
            .touchDown(enabled = enabled) {
                // Press FIRST, then speak — the child sees the squish before the
                // clip has decided whether it is ready.
                motion.press()
                onListen()
            }
            .padding(
                horizontal = ListenPillMetrics.paddingX,
                vertical = ListenPillMetrics.paddingY,
            ),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = text,
            style = Typography.style(
                size = Typography.Size.lg,
                weight = Typography.Weight.bold,
                color = Palette.ink.color,
                ratio = ListenPillMetrics.textRatio,
                fontScale = fontScale,
            ),
        )
    }
}
